subt/ros/robot/src/k2.py

```diff
@@ -0,0 +1,129 @@
+        # calculate backwards what the central joint angle should be.
+        #
+        # Once we know the desired joint angle and current actual joint angle,
+        # we can rotate front and rear halves of the robot in opposite
+        # directions to minimize the difference.
+        #
+        # Rotation of one half of the robot, however, causes pull or push of the
+        # other half, hence we need to make it move forward faster or slower
+        # accordingly.
+
+        # Distance between axes and the middle hinge.
+        # For simplicity, we are neglecting that the front distance (35 cm)
+        # differs from rear distance (37.9 cm).
+        L = 0.3645
+        # The maximum possible angle of the middle joint.
+        MAX_JOINT_ANGLE = math.radians(70)
+
+        try:
+            (_trans, joint_rot) = self.joint_listener.lookupTransform(self.robot_name + '/chassis_back', self.robot_name + '/chassis_front', rospy.Time(0))
+        except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
+            # Not ready yet.
+            return
+
+        current_joint_angle = euler_zyx(joint_rot)[0]
+
+        front_cmd = Twist()
+        rear_cmd = Twist()
+
+        # K2 cannot rotate with a forward/backward velocity.
+        if self.desired.linear.x == 0:
+            self.output_front_publisher.publish(front_cmd)
+            self.output_rear_publisher.publish(rear_cmd)
+            return
+
+        # Initial setting that will be corrected based on joint angle and
+        # desired rotations of each half of the robot.
+        front_cmd.linear.x = self.desired.linear.x
+        rear_cmd.linear.x = self.desired.linear.x
+
+        # Initial rotation needed to keep the robot on the current arc without
+        # slipping.
+        front_cmd.angular.z = rear_cmd.angular.z = (
+                self.desired.linear.x * math.tan(current_joint_angle / 2) / L)
+
+        desired_joint_angle = max(
+                -MAX_JOINT_ANGLE,
+                min(MAX_JOINT_ANGLE,
+                    2 * math.atan2(self.desired.angular.z * L,
+                                   self.desired.linear.x)))
+        #print('current', [math.degrees(alpha) for alpha in euler_zyx(joint_rot)], 'desired', math.degrees(desired_joint_angle), 'because of', math.degrees(self.desired.angular.z))
+
+        P = 5.0  # P control.
+        rot = desired_joint_angle - current_joint_angle
+
+        # To work together, the halfs of the robot need to rotate in opposite
+        # directions.
+        front_cmd.angular.z += P * rot
+        rear_cmd.angular.z += -P * rot
+
+        # Rotating one half of the robot pulls/pushes the other one
+        # forward/backward. To reduce wheel slipping and keep control, we need
+        # to account  for this effect.
+        front_cmd.linear.x += math.sin(current_joint_angle) * rear_cmd.angular.z * L
+        rear_cmd.linear.x += math.sin(current_joint_angle) * front_cmd.angular.z * L
```

Comment:
I am not sure with this part. The `rear_cmd.angular.z`/`front_cmd.angular.z` contains `P * rot` and the robot angular speed computed at line 98. I think the correction (computed here) should be independent to robot angular speed. Corresponding thing is computed [here:](https://github.com/robotika/osgar/blob/master/osgar/drivers/kloubak.py#L172) (differently).

Reply:
The correction totally needs to be a function of the rotation speed of the other half. When, for example, the rear half rotates faster and "pushes", the front half needs to drive faster to avoid slipping and/or bending the middle joint.

I guess your comment touches whether the correction should use *current* rotation velocity, or the *desired* rotation velocity. This code does the later. To be mathematically correct, we should use *current rotation velocity at any given point of time*. At the beginning of the control interval, this is the *current* rotation velocity. At the end of the control interval, it should be the *desired* rotation velocity. If you want, we can settle on a compromise which is an average between the two values, because it should be, on average, more accurate then both of them ;-) We could also increase the control frequency to reduce the difference between the values and thus error that we introduce either at the beginning or at the end of the control interval.

Reply:
I would expect `P * rot` instead of rear_cmd.angular.z there. I am little bit nervous that this calculation is different from my (mentioned above). At least the results should be similar for small angles.

Reply:
OK, this helps me understand what you are saying. (I have troubles matching the other code.)

The difference shows up when `desired_joint_angle = current_joint_angle`, i.e. the robot is already on the desired arc. In that case ... hmm ... let me think a bit more if the correction should be zero or not ...

Reply:
I would merge this PR as it primarily enables Kloubak K2 to be able to run in CloudSim. Feel free to propose changes to 4WD, but they will not be as large as this PR. Also we plan some refactoring (`cloudsim2osgar` vs. `rosmsg`, moving config files to separate folder) which would collide with this PR.